go/launcher/proxy/driverhub/debugger/debugger.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package debugger enables wTL Debugger.
+package debugger
+
+import (
+  "context"
+  "encoding/json"
+  "fmt"
+  "log"
+  "net"
+  "net/http"
+  "os"
+  "regexp"
+  "sync"
+
+  "github.com/bazelbuild/rules_webtesting/go/launcher/errors"
+)
+
+type breakpoint struct {
+  ID        int    `json:"id,omitempty"`
+  Path      string `json:"path,omitempty"`
+  pathRegex *regexp.Regexp
+  Methods   []string `json:"methods,omitempty"`
+}
+
+type command struct {
+  ID         int         `json:"id"`
+  Command    string      `json:"command,omitempty"`
+  Breakpoint *breakpoint `json:"breakpoint,omitempty"`
+}
+
+type request struct {
+  Method string `json:"method,omitempty"`
+  Path   string `json:"path,omitempty"`
+}
+
+type response struct {
+  ID      int     `json:"id,omitempty"`
+  Status  string  `json:"status,omitempty"`
+  Request request `json:"request,omitempty"`
+}
+
+// Debugger is an implementation of the WTL Debugger server.
+type Debugger struct {
+  mu          sync.RWMutex
+  failed      error
+  healthy     bool
+  conn        net.Conn
+  step        bool
+  breakpoints map[int]*breakpoint
+  waiting     chan<- interface{}
+}
+
+func New(port int) *Debugger {
+  d := &Debugger{
+    breakpoints: map[int]*breakpoint{},
+  }
+
+  go d.waitForConnection(port)
+  return d
+}
+
+// Name is the name of this component used in errors and logging.
+func (*Debugger) Name() string {
+  return "WTL Debugger Server"
+}
+
+// Healthy returns nil iff a frontend has connected and has sent a step or continue command.
+func (d *Debugger) Healthy(context.Context) error {
+  d.mu.RLock()
+  defer d.mu.RUnlock()
+
+  if d.failed != nil {
```

Comment:
It's surprising to see this here. The docs for Healthy imply that once the Debugger becomes healthy, it never becomes unhealthy. It's not apparent from the name/type of the `failed` field that it's only ever set by `waitForConnection`. Also, nothing prevents future editors of this code from improperly setting the `failed` field even after the connection is established.

Renaming the field would probably help, but doesn't really eliminate the problem.